File: java/cours/src/session3/TD/MotRecursif/Mot.java
package session3.TD.MotRecursif;

interface Mot {
	String getValue();
	int taille();

	default boolean estVide() {
		return false;
	}

	default boolean estInductif(){
		return false;
	}

	default boolean estConcat(){
		return false;
	}

	// Destructeurs
	default char premiereLettre(){
		throw new UnsupportedOperationException();
	}

	default Mot suite() {
		throw new UnsupportedOperationException();
	}

	default Mot gauche(){
		throw new UnsupportedOperationException();
	}

	default Mot droit(){
		throw new UnsupportedOperationException();
	}

	// Constructeurs
	default Mot vide(){
		return MotVideInductif.VIDE; 
	}

	default Mot motInduct(char n){
		return new MotNonVideInductif(n, this);
	}

	default Mot concat(Mot mot){
		return new MotConcat(this, mot);
	}
}

File: typescript/session3/MotRecursif/Mot.ts
import { Char } from '../Char'
import { MotVideInductif } from './MotVideInductif';
import { MotNonVideInductif } from './MotNonVideInductif';
import { MotConcat } from './MotConcat';

/*
* Mot ::= Vide | cons(char, Mot) | Concatenation(Mot, Mot)
*/
export abstract class Mot {

  abstract getValue(): string;
  abstract taille(): number;

  estVide(): boolean {
    return false;
  }

  estInductif(): boolean {
    return false;
  }

  estCons(): boolean {
    return false;
  }

  estConcat(): boolean {
    return false;
  }

  // Destructeurs
  premiereLettre(): Char {
    throw new Error("Unsupported Exception");
  }

  suite(): Mot {
    throw new Error("Unsupported Exception");
  }

  gauche(): Mot {
    throw new Error("Unsupported Exception");
  }

  droit(): Mot {
    throw new Error("Unsupported Execption");
  }

  // Constructeurs

  vide(): Mot {
    return MotVideInductif.VIDE;
  }

  motInduct(n: Char): Mot {
    return new MotNonVideInductif(n, this);
  }

  concat(mot: Mot): Mot {
    return new MotConcat(this, mot);
  }
}